Improve the error message about too many arguments

Julow opened the pull request:
The "applied to too many arguments" error message confuses me every time I see even though it often has a trivial to fix cause.
The message highlights the function but not the whole application, which can be useful to understand which argument is problematic and where I should add the suggested `;`.

This PR change the location of the main message to span the whole application and adds two sub-messages:
- Where to put the missing `;`, for which the location extends outside the range of the line.
- The extra argument, which helps understand whether it's really about a missing `;` or if I should remove an argument.

The multi-line highlighting is not easy to read, I have a suggestion in https://github.com/ocaml/ocaml/pull/11678

Before:

```
File "./test.ml", line 4, characters 2-13:
4 |   print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
      ^^^^^^^^^^^
Error: This function has type string list -> unit
       It is applied to too many arguments; maybe you forgot a `;'.
```

After:

```
File "./test.ml", lines 4-5, characters 2-15:
4 | ..print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
5 |   print_endline......
Error: This function has type string list -> unit
       It is applied to too many arguments
File "./test.ml", line 4, characters 56-57:
 |   print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
                                                           ^
  Maybe you forgot a ';'.
File "./test.ml", line 5, characters 2-15:
5 |   print_endline "foo"
      ^^^^^^^^^^^^^
  This argument is not expected.
```

Julow committed:
Location: Highlight past the end of the line

Allow to write for example:

    File "./test.ml", line 4, characters 46-47:
     |   f (List.map string_of_int [ 1; 2; 3; 4; 5 ])
                                                     ^
      Maybe you forgot a ';'.

Iterating on the string instead of the range is not useful because range
calculation have already been done.

Julow committed:
Improve the error message about too many arguments

The location of the application, upto the problematic argument, is computed and
used as the location for the main message.
Two sub-messages are added: one for the suggested location of a potential `;`
and the other for highlighting the extra argument.

Before:

    File "./test.ml", line 4, characters 2-13:
    4 |   print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
          ^^^^^^^^^^^
    Error: This function has type string list -> unit
           It is applied to too many arguments; maybe you forgot a `;'.

After:

    File "./test.ml", lines 4-5, characters 2-15:
    4 | ..print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
    5 |   print_endline......
    Error: This function has type string list -> unit
           It is applied to too many arguments
    File "./test.ml", line 4, characters 56-57:
     |   print_lines (List.map string_of_int [ 1; 2; 3; 4; 5 ])
                                                               ^
      Maybe you forgot a ';'.
    File "./test.ml", line 5, characters 2-15:
    5 |   print_endline "foo"
          ^^^^^^^^^^^^^
      This argument is not expected.

Octachron commented on typing/typecore.ml:
```diff
@@ -104,7 +104,9 @@ type error =
-  | Apply_non_function of type_expr
+  | Apply_non_function of type_expr * Location.t * Location.t
+      (** Type of the function, location of extra argument,
+          location of last accepted argument. *)
```
I think that having an inline record with documentation would be even better than having an explanation of what are each tuple elements.

Octachron commented on typing/typecore.ml:
```diff
@@ -5582,13 +5587,30 @@ let report_error ~loc env = function
+              loc_ghost = false }
+          in
+          let sub = [
+            Location.msg ~loc:tail_loc "Maybe you forgot a ';'.";
```
Since this is a suggestion that might be wrong, I would like to have a `Hint` here (seen as "a suggestion that might be wrong") which ends with a question mark:
```
Hint: Did you forget a ';'?
```

Octachron commented on typing/typecore.ml:
```diff
@@ -5582,13 +5587,30 @@ let report_error ~loc env = function
+          in
+          let sub = [
+            Location.msg ~loc:tail_loc "Maybe you forgot a ';'.";
+            Location.msg ~loc:arg_loc "This argument is not expected.";
```
Maybe we could be even more specific: 
```
The function %s cannot be applied to this extra-argument.
```
if we have a identifier at hand and
```
The function cannot be applied to this extra-argument.
```
otherwise?

Julow replied:
That would indeed be better, now that the location is not pointed to the function. I'll investigate how to get that identifier.

Julow replied:
I've implemented this and the other suggestions too.
I've handled other cases than `Texp_ident` for fun and I think this function could evolve to a "describe" function if needed for other error messages. (eg. to returns things like `function f`, `method m`, `module M`)

Julow committed:
Apply_non_function: Print name of function if possible

Increase clarity of the message since the location no longer point
directly to the function.

Introduce 'Typedtree.exp_is_nominal' for testing whether an expression
can be printed as the subject of a sentence with 'Pprintast'.

gasche commented on parsing/location.ml:
```diff
@@ -459,20 +459,20 @@ let highlight_quote ppf
         Format.fprintf ppf "%s | %s@," line_nb line;
         Format.fprintf ppf "%*s   " (String.length line_nb) "";
-        String.iteri (fun i c ->
+        for i = 0 to rightmost.pos_cnum do
```
This change is weird.

I think your intent is to support locations that are outside the source line, "one past" the end. You should say so in a comment, because the code is confusing if we don't know this.

Note that the code computing `lines` above appears to allow "one past" positions, but I wonder if this was intentional or in fact a programming mistake (did the authors realize that their definition of  `end_pos` allow this? I think they believed that there were only including valid line indices in the range). Could you add a comment there as well?

Julow replied:
Agree it's complicated ! Added a comment.

I think the code above works because the end of line has an indice (there's a `pos_cnum` that correspond to it), so only one line can be picked if the location spans it (it won't pick the start of the next line).
I ended up rewriting it in https://github.com/ocaml/ocaml/pull/11678 (with more comments)

Octachron replied:
Wait, `rightmost.pos_cnum` counts the number of character since the start of the file. It is far too large, the right size is `rightmost.pos_cnum - line_cnum_start - 1`.

Julow replied:
Right! That was a big oversight, thanks.

gasche commented on typing/typecore.ml:
```diff
@@ -169,6 +174,34 @@ type error =
+      in
+      parent expr "#" meth'
+  | Texp_instvar (_, _, name) -> Some name.txt
+  | _ -> None
```
I think you can do this with much less code. The problem you have decided to solve is to be able to show the function to the user in the error message, but only if "it can be named", which basically amounts to "it is a short/atomic construct". Is the expression "short" enough that I want to include it in the error message? This is the question.

But you are mixing this the actual printing logic, which is needlessly duplicated. It may also very well be wrong / imperfect; for example, do you handle the case of infix operators properly?

My suggestion: turn this into a `is_short` function that returns a boolean, move it to `typedtree.ml`, and call it from the error message printer (is `funct` short?). For the printing you could call `Untypeast.untype_expression` and then `Pprintast.expression`. (Or maybe you have access to the parsetree directly at error-raising point and don't need `Untypeast`.)

Julow replied:
I wasn't proud of this code! I've implemented it like you suggested and it's much better. I've included more cases and written down why I choose them.

gasche replied:
(I don't currently see the new version in the PR.)

Julow replied:
Sorry, I just pushed it now.

gasche commented on typing/typecore.ml:
```diff
@@ -5582,16 +5619,36 @@ let report_error ~loc env = function
            fprintf ppf "but an expression was expected of type");
-  | Apply_non_function typ ->
-      begin match get_desc typ with
+  | Apply_non_function { func_name; funct; extra_arg_loc; previous_arg_loc } ->
```
I wish `previous_arg_loc` would appear syntactically before `extra_arg_loc` in the code, if it is before in the user source. (This applies in many other places.)

Julow replied:
Wish granted.

gasche commented on typing/typecore.ml:
```diff
@@ -5582,16 +5619,36 @@ let report_error ~loc env = function
+          Location.errorf ~loc:app_loc ~sub
+            "@[<v>@[<2>%a has type@ %a@]\
+             @ It is applied to too many arguments@]"
+            pp_func_name func_name Printtyp.type_expr funct;
```
This is a lot of code, could it (or some of it) move to a helper function?

Julow replied:
It wouldn't be the first error message written that way, so it's entirely out. It's not easy to split further because most of it is computing the locations we want to show specifically for this case, so I kept it as one block.

gasche commented on typing/typecore.mli:
```diff
@@ -154,7 +154,12 @@ type error =
-  | Apply_non_function of type_expr
+  | Apply_non_function of {
+      func_name : string option;
+      funct : type_expr;
```
`func_ty` ? When reading the code I thought `funct` meant `function` and expected it to be an expressions.

Julow replied:
Done.

Octachron commented on parsing/location.ml:
```diff
@@ -459,20 +459,24 @@ let highlight_quote ppf
+        (* Iterate up to [rightmost], which can be bigger than the length of
+           the line because it can span the newline character, which is not
+           included in [line]. Here it is guaranteed that [leftmost] and
+           [rightmost] are on the same line. *)
```
I find this comment slightly confusing: `rightmost` can point to after the end of the line because the `Did you forget a ";"` message points to a location after the last token on the line. The fact that this means pointing to a newline character feels more coincidental than anything.

Maybe:
```
 (* Iterate up to [rightmost], which can be larger than the length of
     the line because we may point to a location after the end of the last token on the line, for instance
      {[
        token
                   ^
       Did you forget ...
       ]}
```
would be clearer?
(I am not sure if the reminder that we are in the single-line case is useful?)

Julow replied:
The reminder about single-line was to make sure that I wasn't accidentally talking about multi-line spans. Your version is better.

Octachron commented on typing/typedtree.ml:
```diff
@@ -845,3 +845,19 @@ let split_pattern pat =
+  | Texp_variant (_, None)
+  | Texp_construct (_, _, [])
+  | Texp_unreachable ->
+      true
```
I would propose to exlude the `Texp_unreachable` because without enough context, it is hard to know that `.` means `unreachable branch`.

Julow replied:
Done.

Octachron commented on typing/typedtree.ml:
```diff
@@ -845,3 +845,19 @@ let split_pattern pat =
+   - Looking like an identifier, small words separated with a '.' or a '#'.
+   - Do not contain spaces when printed.
+   - Look nice as the subject of a sentence, eg. "'Foo.bar' is short." *)
+let rec exp_is_short exp =
```
The name `is_short` gives the impression that it is the size of the expression that matters but we consider that `Tell.Me.O.Muse.Of.That.Ingenious.Hero.Who.Travelled.Far.And.wide` is short. I would rather use your nice description `Look nice as the subject of a sentence`, which could be summarized as `exp_is_nominal`.

Julow replied:
Agree, I used your name and description.

Octachron commented on typing/typedtree.mli:
```diff
@@ -824,3 +824,7 @@ val pat_bound_idents_full:
   computation general_pattern -> pattern option * pattern option
+
+(** Whether an expression is short enough to be printed inline in an error
+    message. *)
```
As stated above,
```
Whether an expression looks nice as the subject of a sentence in a error message
```
is probably a more accurate (and useful) description.

Octachron commented on typing/typecore.ml:
```diff
@@ -5532,6 +5541,36 @@ let report_unification_error ~loc ?sub env err
+    { loc_start;
+      loc_end = { loc_start with pos_cnum = loc_start.pos_cnum + 1 };
+      loc_ghost = false }
+  in
```
There is a corner case here: there is not necessarily a white space after the last valid argument. For instance
```ocaml
let f x y z = x + y + z
let t = f 0 1
    (+)2
```
yields
```
File "test.ml", line 3, characters 7-8:
3 |     (+)2
           ^
  Hint: Did you forget a ';'?
File "test.ml", line 3, characters 7-8:
3 |     (+)2
           ^
  This argument is not expected.
```
One option that might work would be to widen the highlighting interval to include the last character of the previous arg:
```
3 |     (+)2
          ^^
  Hint: Did you forget a ';'?
  ```
  to make it clearer that we are proposing to insert `;` in the middle of the interval.

Julow replied:
Good idea. It's nice when there's a space too, it shows that the semicolon has to be inserted "after the argument".

Octachron commented on typing/typecore.ml:
```diff
@@ -5532,6 +5541,36 @@ let report_unification_error ~loc ?sub env err
+      loc_ghost = false }
+  in
+  let sub = [
+    msg ~loc:tail_loc "Hint: Did you forget a ';'?";
```
We have a new `@{<hint>` tag for hints.

Julow replied:
Nice :)

Octachron commented on typing/typecore.ml:
```diff
@@ -4585,12 +4621,12 @@ and type_application env funct sargs =
   in
   let warned = ref false in
-  let rec type_args args ty_fun ty_fun0 sargs =
+  let rec type_args args ty_fun ty_fun0 sargs prev_arg_loc =
```
Personal nitpicking: as `prev_arg_loc` is a metadata parameter of the functions whereas `sargs` is the decreasing argument in the recursion, I would rather group it with the `args` argument and keep `sargs` as the last argument of the function.

Julow replied:
I paired the location with the argument, which makes the code clearer. However, I'm not sure it decreases the diff size.

Octachron commented on typing/typecore.ml:
```diff
@@ -4640,22 +4676,22 @@ and type_application env funct sargs =
           Some (fun () -> option_none env (instance ty) Location.none)
         in
-        let remaining_sargs, arg =
+        let remaining_sargs, arg, arg_loc =
```
Personal nitpicking 2: the `arg` argument requires some computation whereas `arg_loc` is straightforward to compute. It may make sense to commute those two elements of the tuple so you can decrease the diff size.

Julow replied:
Paired with the `arg` now. The location needs to be passed because `arg` is wrapped in a thunk doing side effects.

Octachron reviewed (approved):
Overall, I like the new information available in the error message. It is a really good idea to point to the supernumerary arguments: it is extremely helpful for beginners and still helpful for the experimented developer.

The hint for `;` could be refined to only appear when adding `;` is correct in the strict mode but this is the existing behavior and improvements are better left to another PR.

Even if I have still have many small remarks, the code is in good shape and could be merged, once the missing hint tag is added.

Octachron commented on typing/typecore.ml:
```diff
@@ -5532,6 +5541,36 @@ let report_unification_error ~loc ?sub env err
+  in
+  let sub = [
+    msg ~loc:tail_loc "Hint: Did you forget a ';'?";
+    msg ~loc:extra_arg_loc "This argument is not expected.";
```
Note that the functor error message uses `This extra argument` to emphasize that the argument doesn't fit the function type. I think could reuse the term here.

Julow replied:
Done.

Julow committed:
Missing semicolon hint: Span the end of the argument

Extend the location to the right to cover the end of the previous
argument as well.
This renders well when there's no space between the two arguments and
shows better that the semicolon has to be inserted (between two
position).

Julow commented:
Ready for review again. I rebased and cleaned the history.

> The hint for ; could be refined to only appear when adding ; is correct in the strict mode but this is the existing behavior and improvements are better left to another PR.

Added to my TODO.

Octachron commented on typing/typecore.ml:
```diff
@@ -4528,16 +4533,25 @@ and type_application env funct sargs =
+                | (_, None) :: hd -> previous_arg_loc hd
+                | [] -> funct.exp_loc
+              in
+              let previous_arg_loc = previous_arg_loc typed_args in
```
This is:
```ocaml
 typed_args 
 |> List.find_map (function (_,Some (_,loc)) -> loc | _ -> None )
 |> Option.value ~default:funct.exp_loc
 ```

Julow replied:
Better :)

Octachron commented on typing/typecore.ml:
```diff
@@ -5619,6 +5632,40 @@ let report_unification_error ~loc ?sub env err
+       of the argument. *)
+    let arg_end = previous_arg_loc.loc_end in
+    { loc_start = cnum_offset ~-1 arg_end;
+      loc_end = cnum_offset ~+1 arg_end;
```
This use of unary `+` is a bit surprising (this is the only instance in the whole compiler)?

Julow replied:
That's nice, isn't it ?

Octachron commented on typing/typecore.ml:
```diff
@@ -4643,11 +4656,11 @@ and type_application env funct sargs =
                 sargs,
                 if optional && List.mem_assoc Nolabel sargs then
-                  eliminate_optional_arg ()
+                  Some (eliminate_optional_arg (), Location.none)
```
Using `Location.none` here is valid due to a non-local invariant: we are only using the location of the last argument and the test above  (`List.mem_assoc Nolabel sargs`) guarantees that there are more arguments after the current one. I would recommend to at least add a comment that we can use `Location.none` here because we know that there would be at least one argument with a valid location in `sargs`.

Julow replied:
I wasn't sure how to define that clearly so I switching to using an option. Not all argument have a location.

Octachron commented on parsing/location.ml:
```diff
@@ -459,20 +459,28 @@ let highlight_quote ppf
             Format.fprintf ppf "@{<%s>" highlight_tag;
           if ISet.mem iset ~pos then Format.pp_print_char ppf '^'
-          else if pos < rightmost.pos_cnum then begin
+          else if pos < rightmost.pos_cnum && i < String.length line then begin
```
The first test `pos < rightmost.pos_cnum`  is redundant once the lower bound of the `for` loop is correct.

Julow replied:
Fixed, thanks.

Octachron commented on tools/check-typo-since:
```diff
@@ -32,6 +32,6 @@ check_typo_since() {
 case $# in
     0) echo "usage: check-typo-since <git reference>"; exit 2;;
-    1) check_typo_since $1; break;;
+    1) check_typo_since $1;;
```
This is not a syntax error ? And this fails silently with `sh` or `dash` but `bash` emits an error. It is probably better to split this commit in a separate PR that I can approve shortly.

Julow replied:
Moved to https://github.com/ocaml/ocaml/pull/11914

Octachron commented on typing/typedtree.ml:
```diff
@@ -844,3 +844,18 @@ let split_pattern pat =
   in
   split_pattern pat
+
+(* Expressions are considered short if they satisfy one of:
```
short ⇒ nominal

Octachron commented on typing/typedtree.ml:
```diff
@@ -844,3 +844,18 @@ let split_pattern pat =
+
+   - Similar to an identifier: small words separated by '.' or '#'.
+   - Do not contain spaces when printed.
+   - Can be used as the subject of a sentence, eg. "'Foo.bar' is short." *)
```
short ⇒ nominal . You could also restructure the comments to make this last condition the general summary:

```
(* Expressions are considered nominal if they can be used as the subject of a sentence or action.
    In practice, we consider that an expression is nominal if they satisfy one of:
   - Similar to an identifier:  words separated by '.' or '#'.
   - Do not contain spaces when printed.
  *)
```

Julow replied:
Updated, thanks.

Octachron commented:
This is good to merge once the check-typo commit has been split off.

Octachron commented:
Merged, thanks a lot for the great improvement to the compiler error messages!
